### DISCUSSION TITLE
Some questions in bases

### DISCUSSION THREAD

**Discussion by Kardiem** [ANLAB-KAIST/KENSv3#91] [Project Q&A 3. Internet Layer (IP/Routing)]
I've started this assignment quite late, but i need to understand some conflicts about this project.

1. I've understand that 'packetarrived()' function is for receive the packet and do proper action, and 'timercallback()' does same time measuring. But then what does 'ripquerry()' function do?

i'm confused with the range of 'ripquery()' function do at first place between following 3 cases.
1) initiallizing the routing graph, and sending the first broadcast of the router which called this function.
2) initiallize has already done, just send the first broadcast.
3) initialize and routing has already done from outside of function, ripquery just cast the result.

2. Since there was no routing table structure in hpp file. Do we have to make an new routing table map? if so, how do we initialize it? (initializing its cost to 0 is easy, but how do we get the edges of routing graph from test case?)
I wonder there's an already existing structure or map to help to find the neighbor routers or the edges of the entire graph.

3.  RIP format has 1 IP address and 1 metric. So then if the response should send more than 2 routing information, then the router should send the RIP format packet more than once?

**Comment by Pusnow** [comment]
> Q: I've understand that 'packetarrived()' function is for receive the packet and do proper action, and 'timercallback()' does same time measuring. But then what does 'ripquerry()' function do?

It is used for grading. Only grading script will call it.

(Below is PA4's document)

Section 4 of RFC1058 describes administrative controls.  In this assignment, you should implement only one control function used to assess correctness of your protocol implementation: ripQuery().

```c++
Size RoutingAssignment::ripQuery(const ipv4_t &ipv4);
```

The ripQuery() call receives an IP address as a parameter and should return the cost to reach the IP address. In RIPv1, the cost is just the hop-count.


> Q: Since there was no routing table structure in hpp file. Do we have to make an new routing table map? if so, how do we initialize it? (initializing its cost to 0 is easy, but how do we get the edges of routing graph from test case?)
I wonder there's an already existing structure or map to help to find the neighbor routers or the edges of the entire graph.

You have to make a new table. Initialize it like below. You can use `getPortCount()` and `getIPAddr()` to obtain IP addresses of interfaces. https://github.com/ANLAB-KAIST/KENSv3/wiki/Tip:-Figuring-out-the-source-IP-address-to-use

(Below is PA4's document)

Given a topology during initialization, every node starts with itself in the distance vector table with a metric of 0. If a node has multiple interfaces, it enters all the IP addresses of the interfaces to the distance vector table.


> Q: RIP format has 1 IP address and 1 metric. So then if the response should send more than 2 routing information, then the router should send the RIP format packet more than once?

According to RFC1058, you can repeat the datagram from address family identifier through metric up to 25 times.

https://github.com/ANLAB-KAIST/KENSv3/blob/b907a1ccca4663a94f9e5610a76a0849647ff033/Documentation/rfc/rfc1058.txt#L1040-L1044

**Reply by Kardiem** [reply]
1. so then ripQuery does the start of the routing algorithm and initializing the routing table? (which means broadcast the request to every routers?)

(설명이 부족한것 같아서 한국말로 하면, ripQuery가 테이블 초기화 -> 라우팅 알고리즘 시작 -> (packetarrived를 통해 자연스레 알고리즘 끝까지 진행) -> 완료되면 table에서 값을 찾아서 반환 이렇게 3가지를 진행하는건가요?)

2. so then routing table has every edge on the table at first place, then the routing table is updated with response from neighbor routers?

**Reply by Pusnow** [reply]
> Q: so then ripQuery does the start of the routing algorithm and initializing the routing table? (which means broadcast the request to every routers?)

No, the `ripQuery` simply returns current cost in the table. It does not initiate routing algorithm or send packet to other routers.

> Q: so then routing table has every edge on the table at first place, then the routing table is updated with response from neighbor routers?

Initially, the routing table only has edges of its node, then the table is updated and added with response from neighbor routers.
